### PR TITLE
DS918+/7.0-41890 with virtio

### PR DESCRIPTION
**Pull request by jumkey**

### CODE REVIEW

**Comment by jumkey** [comment]
can work with this patch [via](https://xpenology.com/forum/topic/45795-redpill-the-new-loader-for-624-discussion/?do=findComment&comment=209617)
```patch
diff --git a/build-loader.sh b/build-loader.sh
index a5288b6..eed834c 100755
--- a/build-loader.sh
+++ b/build-loader.sh
@@ -264,6 +264,8 @@ readonly BRP_RDMOD_NAME="rd.gz" # name of the ramdisk in the final image
 brp_cp_flat "${BRP_ZLINUX_PATCHED_FILE}" "${BRP_OUT_P1}/${BRP_ZLINMOD_NAME}"
 brp_cp_flat "${BRP_RD_REPACK}" "${BRP_OUT_P1}/${BRP_RDMOD_NAME}"
 brp_cp_flat "${BRP_TMP_GRUB_CONF}" "${BRP_OUT_P1}/boot/grub/grub.cfg"
+# create SynoBootLoader.conf if not exsit for 7.0
+mkdir -p ${BRP_OUT_P1}/EFI/boot; echo -n "" >> ${BRP_OUT_P1}/EFI/boot/SynoBootLoader.conf
 pr_process_ok

 ##### CLEANUP ##########################################################################################################
```

**Comment by semool** [comment]
> can work with this patch [via](https://xpenology.com/forum/topic/45795-redpill-the-new-loader-for-624-discussion/?do=findComment&comment=209617)
> 
> ```diff
> diff --git a/build-loader.sh b/build-loader.sh
> index a5288b6..eed834c 100755
> --- a/build-loader.sh
> +++ b/build-loader.sh
> @@ -264,6 +264,8 @@ readonly BRP_RDMOD_NAME="rd.gz" # name of the ramdisk in the final image
>  brp_cp_flat "${BRP_ZLINUX_PATCHED_FILE}" "${BRP_OUT_P1}/${BRP_ZLINMOD_NAME}"
>  brp_cp_flat "${BRP_RD_REPACK}" "${BRP_OUT_P1}/${BRP_RDMOD_NAME}"
>  brp_cp_flat "${BRP_TMP_GRUB_CONF}" "${BRP_OUT_P1}/boot/grub/grub.cfg"
> +# create SynoBootLoader.conf if not exsit for 7.0
> +mkdir -p ${BRP_OUT_P1}/EFI/boot; echo -n "" >> ${BRP_OUT_P1}/EFI/boot/SynoBootLoader.conf
>  pr_process_ok
> 
>  ##### CLEANUP ##########################################################################################################
> ```

You can also insert the dirs and the file direct in the img Template

File: config/DS918+/7.0-41890/ramdisk-002-init-script.patch [hunk 1]
```diff
@@ -0,0 +1,28 @@
+--- a/linuxrc.syno.impl
```
Comment (ttg-public) [inline config/DS918+/7.0-41890/ramdisk-002-init-script.patch:1]: Shouldn't this be a symlink instead? By looking at this it seems that the patch is the same.

File: config/DS918+/7.0-41890/ramdisk-003-post-init-script.patch [hunk 2]
```diff
@@ -0,0 +1,30 @@
+--- usr/sbin/init.post
```
Comment (ttg-public) [inline config/DS918+/7.0-41890/ramdisk-003-post-init-script.patch:1]: Shouldn't this one be a symlink as well?

**Review by ttg-public — COMMENTED** [review]
Thank you for the contribution!

We left two comments to the files. Additionally, you should remove some of the changed chmods as they shouldn't be in the diff ;)

The biggest question is what SUPPOSE to be in that `EFI` folder? Do you maybe have a physical 918 or similar to check? It's may also be somewhere in the PAT (we didn't check that yet ;)).

File: config/DS918+/7.0-41890/ramdisk-002-init-script.patch (hunk 1 above)
Comment (semool) [reply config/DS918+/7.0-41890/ramdisk-002-init-script.patch:1]: no thats changes in the files in DSM7.

File: config/DS918+/7.0-41890/ramdisk-003-post-init-script.patch (hunk 2 above)
Comment (semool) [reply config/DS918+/7.0-41890/ramdisk-003-post-init-script.patch:1]: same as by the impl patch. Some changes in DSM7.

**Comment by jumkey** [comment]
as @semool commented
this two files changes in DSM7

mntroot from DS1515+ @ttg-public 
```console
root@debian:~/syno/mntroot# ls -l -R
.:
total 8784
drwxr-xr-x 3 root root    4096 Aug  5 12:34 EFI
-rw-r--r-- 1 root root     100 Jul 30 14:07 GRUB_VER
-rw-r--r-- 1 root root 6055885 Jul 30 14:07 rd.gz
-rw-r--r-- 1 root root 2924624 Jul 30 14:07 zImage

./EFI:
total 4
drwxr-xr-x 2 root root 4096 Aug  5 12:34 boot

./EFI/boot:
total 208
-rw-r--r-- 1 root root    468 Jul 30 14:07 SynoBootLoader.conf
-rw-r--r-- 1 root root 205616 Jul 30 14:07 SynoBootLoader.efi
root@debian:~/syno/mntroot# cat EFI/boot/SynoBootLoader.conf
serial --unit=1 --speed=115200
terminal serial
default 1
timeout 3
verbose
hiddenmenu
fallback 0

title SYNOLOGY_1
        root (hd0,0)
        kernel /zImage root=/dev/md0 ihd_num=0 netif_num=4 pwrctl_pin=N0754N0850 DiskIdxMap=020006 syno_hw_version=DS1515+
        initrd /rd.gz

title SYNOLOGY_2
        root (hd0,1)
        cksum /grub_cksum.syno
        vender /vender show
        kernel /zImage root=/dev/md0 ihd_num=0 netif_num=4 pwrctl_pin=N0754N0850 DiskIdxMap=020006 syno_hw_version=DS1515+
        initrd /rd.gz
root@debian:~/syno/mntroot#
```

**Comment by ttg-public** [comment]
@jumkey Which platform did you use for testing? We were attempting to run v7 with 918+ on Proxmox 6 as well as VBox 6 and in both cases the USB isn't working properly. What's worse it seems like we stepped into the same bug as with 3617 where USB driver is loaded but no devices are detected. In fact no hot-plug capable devices are working. Despite VirtIO drivers being loaded the ethernet isn't detected either.

Of course due to that we didn't even try installing as error 13 is unavoidable due to missing `/dev/synoboot`.

**Comment by jumkey** [comment]
> @jumkey Which platform did you use for testing? We were attempting to run v7 with 918+ on Proxmox 6 as well as VBox 6 and in both cases the USB isn't working properly. What's worse it seems like we stepped into the same bug as with 3617 where USB driver is loaded but no devices are detected. In fact no hot-plug capable devices are working. Despite VirtIO drivers being loaded the ethernet isn't detected either.
> 
> Of course due to that we didn't even try installing as error 13 is unavoidable due to missing `/dev/synoboot`.

@ttg-public tested on Proxmox 7 with conf
```console
args: -device 'qemu-xhci,id=usb-ctl-synoboot,addr=0x18' -drive 'id=usb-drv-synoboot,file=/var/lib/vz/images/103/synoboot.img,if=none,format=raw' -device 'usb-storage,id=usb-stor-synoboot,drive=usb-drv-synoboot,bootindex=1,removable=off'
boot:
cores: 2
memory: 2048
name: ds7
net0: virtio=8A:02:1B:**:**:**,bridge=vmbr0,firewall=1
net1: virtio=92:46:4B:**:**:**,bridge=vmbr0,firewall=1
numa: 0
ostype: l26
sata0: local-lvm:vm-103-disk-1,size=32G
scsihw: virtio-scsi-pci
serial0: socket
serial1: socket
serial2: socket
serial3: socket
smbios1: uuid=********-db35-425f-992a-76cd30296558
sockets: 1
vmgenid: ********-63d3-4c24-9228-6119780254d4
```

**Comment by ttg-public** [comment]
Let's just call it... magic - it worked with the config you posted on Proxmox 6 despite being really no different than ours (we use loop device for USB to make switching easier & q35 machine type but that shouldn't be a problem...).

We merged the changes but since we were in the middle of reorganizing symlinks hell (they're gone!) we had to change the PR and split it a bit ;) See https://github.com/RedPill-TTG/redpill-load/compare/ded0ac6375d441a86dcdfee2ec399d95225a12db...cfd70c78cc231709a2f79439e09eb3941b94b83a - of course we credited you on it.

Thank you so much for the contribution and the patience :)

**Comment by LS12tthappy** [comment]
> Let's just call it... magic - it worked with the config you posted on Proxmox 6 despite being really no different than ours (we use loop device for USB to make switching easier & q35 machine type but that shouldn't be a problem...).
> 
> We merged the changes but since we were in the middle of reorganizing symlinks hell (they're gone!) we had to change the PR and split it a bit ;) See [ded0ac6...cfd70c7](https://github.com/RedPill-TTG/redpill-load/compare/ded0ac6375d441a86dcdfee2ec399d95225a12db...cfd70c78cc231709a2f79439e09eb3941b94b83a) - of course we credited you on it.
> 
> Thank you so much for the contribution and the patience :)

I guess the key is the three **serial** added (?

**Pull request closed without merging**